### DISCUSSION TITLE
python 3.12

### DISCUSSION THREAD

**Discussion by thezoggy** [sabnzbd/sabnzbd#2701] [General]
py3.12 officially out today, setup new pyenv and went through motions to run latest dev + tests on my windows 10 x64 box.
```
python -m pip install --upgrade pip wheel
pip3 install --upgrade -r requirements.txt
pip3 install --upgrade -r builder/requirements.txt
python tools/make_mo.py
```
then on installing our tests requirements,
` pip3 install --upgrade -r tests/requirements.txt`

`rpds-py` currently fails to install, missing wheel
```
Collecting rpds-py>=0.7.1 (from jsonschema<5,>=3.2.0->tavern>=2.0.0->-r tests/requirements.txt (line 11))
  Using cached rpds_py-0.10.3.tar.gz (17 kB)
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Preparing metadata (pyproject.toml) ... error
  error: subprocess-exited-with-error

  × Preparing metadata (pyproject.toml) did not run successfully.
  │ exit code: 1
  ╰─> [6 lines of output]

      Cargo, the Rust package manager, is not installed or is not on PATH.
      This package requires Rust and Cargo to compile extensions. Install it through
      the system's package manager or via https://rustup.rs/

      Checking for Rust toolchain....
      [end of output]
```
already reported at: https://github.com/crate-py/rpds/issues/26

--

looks like everything working fine, test nzb was good. nothing showing up in logs differently.

**Comment by sanderjo** [comment]
Cool! Compliments to SABnzbd and the python eco systems that it all works.

I tried with a docker image with python3.12, and that works too.
```
docker run -it python:3.12-bookworm /bin/bash
```


One note:

Within that debian python3.12 docker image:
```
# /usr/bin/env python3
Python 3.12.0 (main, Oct  3 2023, 01:48:15) [GCC 12.2.0] on linux

```
good, but:

```
# ls -al /usr/bin/python3
lrwxrwxrwx 1 root root 10 Apr  9 10:22 /usr/bin/python3 -> python3.11
```
so `/usr/bin/python3` (as hard-specced in the first line of SABnzbd.py) is not pointing to python3.12, but python3.11. So I had to start sab with `python3.12 SABnzbd.py` not `./SABnzbd.py`.

**Comment by Safihre** [comment]
Yes I've been testing 3.12 since the beta's. Resolved al our deprecation warnings, but there's a few in the depencies.

 For sabctools we already build a wheel for 3.12 since a few months 🤗
